mmf.c

```diff
@@ -0,0 +1,220 @@
+#include <stdio.h>
+#include <sys/mman.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <getopt.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include "mmf.h"
+#include "mmf-mmap.h"
+#include "mmf-shm.h"
+
+
+/* for_each_map - iterate over a test_t's map[] array
+ *   t: test_t to span
+ *   m: this iteration's map_t *
+ *   o: allocation order, used to calculate array index
+ */
+#define for_each_map(t, m, o) \
+        for ((o) = (t).min_order, (m) = &(t).mappings[(o) - (t).min_order]; \
+             (o) <= (t).max_order; \
+             (o)++, (m) = &(t).mappings[(o) - (t).min_order])
+
+int do_mmap = 0;
+int do_shmem = 0;
+int do_fork = 0;
+int demo = 0;
+int debug_print = 0;
+
+unsigned long long map_size = 1UL << 12;
+
+void usage(){
+	printf("Usage: ./mmap+memset+fork [OPTION]\n");
+	printf("Tests virtual adress space updates with fork and mmap options\n\n");
+	printf("-m, --mmap		allocates memory using mmap\n");
+	printf("-c, --shmem		allocates memory using shared memory\n");
+	printf("-d, --debug		prints out the PID and address touched by memset\n");
+	printf("-k, --keystroke		same as debug, but requires a char to be entered to show each line\n");
+	printf("-h, --help              prints out help message\n");
+	printf("-f, --fork_child        causes program to fork a child process\n");
+	printf("-b, --begin             mmaps and memsets from 2^(arg)\n");
+	printf("-e, --end               mmaps and memsets till 2^(arg)\n");
+	printf("-o, --mmap_order        mmaps given 2^(arg) size\n");
```

Comment:
nitpick: `-b` `-e-` and `-o` should say "maps" and not "mmaps" since this program now supports multiple mapping methods.